src/umode.rs

```diff
@@ -0,0 +1,205 @@
+// Copyright (c) 2022 by Rivos Inc.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::smp::PerCpu;
+use crate::umode_mem::UmodeMemoryRange;
+
+use arrayvec::ArrayVec;
+use core::cell::{RefCell, RefMut};
+use riscv_elf::{ElfMap, ElfSegmentPerms};
+use riscv_pages::PageSize;
+use riscv_regs::GeneralPurposeRegisters;
+use s_mode_utils::print::*;
+use spin::Once;
+
+/// Host GPR and which must be saved/restored when entering/exiting U-mode.
+#[derive(Default)]
+#[repr(C)]
+struct HostCpuRegs {
+    gprs: GeneralPurposeRegisters,
+    stvec: u64,
+    sscratch: u64,
+}
+
+/// Umode GPR and CSR state which must be saved/restored when exiting/entering U-mode.
+#[derive(Default)]
+#[repr(C)]
+struct UmodeCpuRegs {
+    gprs: GeneralPurposeRegisters,
+    sepc: u64,
+}
+
+/// CSRs written on an exit from virtualization that are used by the host to determine the cause of
+/// the trap.
+#[derive(Default)]
+#[repr(C)]
+struct TrapRegs {
+    scause: u64,
+    stval: u64,
+}
+
+/// CPU register state that must be saved or restored when entering/exiting U-mode.
+#[derive(Default)]
+#[repr(C)]
+struct UmodeCpuArchState {
+    host_regs: HostCpuRegs,
+    umode_regs: UmodeCpuRegs,
+    trap_csrs: TrapRegs,
+}
+
+/// Errors returned by U-mode runs.
+#[derive(Debug)]
+pub enum Error {
+    /// ELF segment out of range,
+    InvalidElf,
+    /// MAX_RESET_ARGS is too smal for ELF file.
+    ResetVectorFull,
+    /// Task already active.
+    TaskBusy,
+}
+
+#[derive(Debug)]
+// Holds information on how to resets an U-mode memory area to its original state.
+struct UmodeResetArea {
+    // The range of memory to be reset.
+    range: UmodeMemoryRange,
+    // Data to be copied at the beginning of the area.
+    data: Option<&'static [u8]>,
+}
+
+impl UmodeResetArea {
+    fn reset(&self) {
+        let range = &self.range;
+        let mut copied = 0;
+        // Copy data at the beginning.
+        if let Some(data) = self.data {
+            self.range.copy_to_umode(data);
+            copied = data.len();
+        }
+        // If there is still space available after the copy, zero the rest.
+        if copied < range.len() {
+            // Unwrap okay. Offset is valid because `copied` is smaller than the range length.
+            range.offset(copied).unwrap().clear();
+        }
+    }
+}
+
+// Global U-mode state.
+static UMODE: Once<Umode> = Once::new();
+
+// Maximum number of reset areas.
+const MAX_RESET_AREAS: usize = 2;
+
+/// Represents the U-mode global state.
+pub struct Umode {
+    entry: u64,
+    reset_areas: ArrayVec<UmodeResetArea, MAX_RESET_AREAS>,
+}
```

Comment:
Isn't this information already available to us in `HypMap` and `ElfMap`? Do we need to duplicate it here?

Reply:
Yes, you hinted this at the previous iteration. `entry` is definitely there and doesn't have to be copied. Creating the reset areas (a small numbers) help with some basic sanity check and preparation of data for having simpler reset code. I want to avoid iterating over all ELF segments, selecting the writable ones, rounding up the page of the size and cutting data to size every time.

Reply:
I think a word about the current design that might have been missed is necessary for context.

ELFs can be a pain to parse. Not because the format is complicated, but because there are a lot of things to check every time we want to use it in a safe way. Every offset, every value. The way I went in `riscv-elf` is to create a simple library that does all these checks once, creating a data structure that is essentially a sanitized image of the ELF file parts we're interested in, easy to use from the salus code. While the data is references is in the ELF file directly, this structure is not static. Rather, gets consumed later on.

This data structure, `ElfMap`, is then passed around to more specialized code (`HypMap` and `Umode`) to create the data it needs. The advantage of that is that `ElfMap` can be maintained simple and generic, and the other code can just parse this information to something it's much easier to handle. More importantly, this structure is consumed, and dropped by the last place that uses it (was `HypMap`, here becomes `umode`).

Now, based on the whole `unsafe` discussion, I am going to remove `UmodeMemRange`, which removes the need to create specialized reset areas structures in Umode. Furthermore, I agree that having a single point for mapping information is much better than disseminating it across the code.

In my next iteration, the direction I am thinking of going is to have `HypMap` being able to return mapping information it got from the `ElfMap`, and use it in umode on reset.

Reply:
Reset not implemented in this iteration. But most likely still going with what I said above.